### PR TITLE
add test for pipelined pessimistic locking

### PR DESCRIPTION
**Pull request by youjiali1995**
### What problem does this PR solve? 
Test the success rate of pipelined pessimistic locking.

### What is changed and how does it work?
Add a new test:
A pessimistic transaction using pipelined pessimistic locking can fail to commit when
  1. Fails to write pessimistic locks due to region error;
  2. And fails to amend pessimistic locks when prewrites if forUpdateTs is updated due to write conflict.

This test creates a table with tableSize rows. Each row has contenderCount pessimistic transactions updating it concurrently and
schedule regions of the table so that pipelined pessimistic lock can fail.

### Check List 

Tests 
- Manual test (add detailed scripts or steps below)

Code changes
 - Has Go code change

### Does this PR introduce a user-facing change?:

```release-note
NONE
```

### CODE REVIEW

File: tests/pessimistic/pipeline/pipeline.go [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+	return err
+}
+
+func (c *pipelineClient) checkKeyIsLocked(ctx context.Context, row int) (bool, error) {
+	tx, err := c.db.Begin()
+	if err != nil {
+		return false, err
+	}
+	defer tx.Rollback()
+	rows, err := tx.QueryContext(ctx, fmt.Sprintf(`SELECT * FROM %s WHERE id in (%d, %d) FOR UPDATE NOWAIT`, tableName, row, row+1))
```
Comment (cosven) [inline tests/pessimistic/pipeline/pipeline.go:228]: what does the `row + 1` means here? 
I think it should be `row + c.TableSize`?

File: tests/pessimistic/pipeline/pipeline.go [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+		default:
+		}
+
+		tx, err := c.db.Begin()
+		if err != nil {
+			log.Fatalf("fail to begin a transaction: %v", err)
+		}
+		// When pipelined pessimistic locking is enabled, returning successful result means the lock request has been proposed.
+		if _, err := tx.ExecContext(ctx, fmt.Sprintf(`UPDATE %s SET v = v - 1 WHERE id = %d`, tableName, row)); err != nil {
+			log.Fatalf("fail to update: %v", err)
```
Comment (cosven) [inline tests/pessimistic/pipeline/pipeline.go:255]: I'm wondering if panic is suitable here. 

Usaually, panic means that the test case is failed and there maybe a bug related to **pipelined lock**.  However, we can't infer the conclusion from this error.

File: tests/pessimistic/pipeline/pipeline.go [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+// DumpState the database state(also the model's state)
+func (c *pipelineClient) DumpState(ctx context.Context) (interface{}, error) {
+	panic("not implemented")
+}
+
+// Start runs self scheduled cases
+// this function will block Invoke trigger
+// if you want to schedule cases by yourself, use this function only
+func (c *pipelineClient) Start(ctx context.Context, cfg interface{}, clientNodes []cluster.ClientNode) error {
+	c.scatterTable(dbName, tableName)
```
Comment (cosven) [inline tests/pessimistic/pipeline/pipeline.go:177]: To my understanding, the table `t` only has one region. Will the scatter table works as we expect?

File: tests/pessimistic/pipeline/pipeline.go (hunk 2 above)
Comment (cosven) [reply tests/pessimistic/pipeline/pipeline.go:255]: If table `t` has multiple regions, I think we can enable `random-merge-scheduler` in nemesis.

File: argo/workflow/pipelined-pessimistic-locking.yaml [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+      - name: ns
+        value: tipocket-pipelined-pessimistic-locking
+      - name: purge
+        value: "true"
+      - name: image-version
+        value: release-4.0-nightly
+      - name: storage-class
+        value: local-storage
+      - name: nemesis
+        value: ""
```
Comment (cosven) [inline argo/workflow/pipelined-pessimistic-locking.yaml:17]: how about enable `shuffle region scheduler` and `random kill`?

**Comment by cosven** [comment]
I run this test case in local and it reports the following error
```
~/c/tipocket youjiali1995-pipelined-pessimistic-locking* > bin/pipelined-locking -tidb-server 127.0.0.1:4000 -pd-server 127.0.0.1:2379 -tikv-server 127.0.0.1:20160 -tikv-server 127.0.0.1:20161 -tikv-server 127.0.0.1:20162 -run-time 100s  -local-mode=true -strict=true
2020/11/10 08:41:42 fixture.go:237: [info] Git Commit Hash: 6fd298ac05a546e6e325f167a342d28a8b320fb9
2020/11/10 08:41:42 fixture.go:238: [info] UTC Build Time: 2020-11-10 08:29:02
W1110 08:41:42.012155   14314 client_config.go:541] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
W1110 08:41:42.012193   14314 client_config.go:546] error creating inClusterConfig, falling back to default config: unable to load in-cluster configuration, KUBERNETES_SERVICE_HOST and KUBERNETES_SERVICE_PORT must be defined
2020/11/10 08:41:42 client.go:109: [warning] build config failed: invalid configuration: no configuration has been provided
2020/11/10 08:41:42 client.go:47: [warning] error creating kube-client: must provide non-nil rest.Config to client.New
2020/11/10 08:41:42 suit.go:87: [info] deploy cluster success, node:[ 127.0.0.1:4000  127.0.0.1:20160  127.0.0.1:20161  127.0.0.1:20162  127.0.0.1:2379], client node:[ 127.0.0.1:4000]
2020/11/10 08:41:42 control.go:99: [info] start controller with &{Mode:Self scheduled mode DB: Nodes:[ 127.0.0.1:4000  127.0.0.1:20160  127.0.0.1:20161  127.0.0.1:20162  127.0.0.1:2379] ClientNodes:[ 127.0.0.1:4000] ClientCount:1 ChaosNamespace: RunRound:1 RunTime:1m40s RequestCount:0 History: ClientConfig:<nil>}
2020/11/10 08:41:42 control.go:348: [info] begin to set up database
2020/11/10 08:41:42 control.go:350: [info] begin to set up database on  127.0.0.1:2379
2020/11/10 08:41:42 control.go:350: [info] begin to set up database on  127.0.0.1:4000
2020/11/10 08:41:42 control.go:350: [info] begin to set up database on  127.0.0.1:20160
2020/11/10 08:41:42 control.go:350: [info] begin to set up database on  127.0.0.1:20161
2020/11/10 08:41:42 control.go:350: [info] begin to set up database on  127.0.0.1:20162
2020/11/10 08:41:42 control.go:371: [info] begin to set up client
2020/11/10 08:41:42 control.go:374: [info] begin to set up db client for node  127.0.0.1:4000
2020/11/10 08:41:42 panic_check.go:34: [warning] plugin panic check won't work
2020/11/10 08:41:42 control.go:298: [info] run client 0...
2020/11/10 08:41:45 leader_shuffle.go:269: [info] [leader shuffler] [leader=4] Merge leader region #126 and #2
2020/11/10 08:41:46 pipeline.go:264: [fatal] fail to lock: <nil>
```

File: tests/pessimistic/pipeline/pipeline.go (hunk 1 above)
Comment (youjiali1995) [reply tests/pessimistic/pipeline/pipeline.go:228]: Yes. I missed it.

File: tests/pessimistic/pipeline/pipeline.go (hunk 3 above)
Comment (youjiali1995) [reply tests/pessimistic/pipeline/pipeline.go:177]: Because of the leader shuffler, it will have more than one regions.

**Comment by youjiali1995** [comment]
> I run this test case in local and it reports the following error
> 
> ```
> ~/c/tipocket youjiali1995-pipelined-pessimistic-locking* > bin/pipelined-locking -tidb-server 127.0.0.1:4000 -pd-server 127.0.0.1:2379 -tikv-server 127.0.0.1:20160 -tikv-server 127.0.0.1:20161 -tikv-server 127.0.0.1:20162 -run-time 100s  -local-mode=true -strict=true
> 2020/11/10 08:41:42 fixture.go:237: [info] Git Commit Hash: 6fd298ac05a546e6e325f167a342d28a8b320fb9
> 2020/11/10 08:41:42 fixture.go:238: [info] UTC Build Time: 2020-11-10 08:29:02
> W1110 08:41:42.012155   14314 client_config.go:541] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
> W1110 08:41:42.012193   14314 client_config.go:546] error creating inClusterConfig, falling back to default config: unable to load in-cluster configuration, KUBERNETES_SERVICE_HOST and KUBERNETES_SERVICE_PORT must be defined
> 2020/11/10 08:41:42 client.go:109: [warning] build config failed: invalid configuration: no configuration has been provided
> 2020/11/10 08:41:42 client.go:47: [warning] error creating kube-client: must provide non-nil rest.Config to client.New
> 2020/11/10 08:41:42 suit.go:87: [info] deploy cluster success, node:[ 127.0.0.1:4000  127.0.0.1:20160  127.0.0.1:20161  127.0.0.1:20162  127.0.0.1:2379], client node:[ 127.0.0.1:4000]
> 2020/11/10 08:41:42 control.go:99: [info] start controller with &{Mode:Self scheduled mode DB: Nodes:[ 127.0.0.1:4000  127.0.0.1:20160  127.0.0.1:20161  127.0.0.1:20162  127.0.0.1:2379] ClientNodes:[ 127.0.0.1:4000] ClientCount:1 ChaosNamespace: RunRound:1 RunTime:1m40s RequestCount:0 History: ClientConfig:<nil>}
> 2020/11/10 08:41:42 control.go:348: [info] begin to set up database
> 2020/11/10 08:41:42 control.go:350: [info] begin to set up database on  127.0.0.1:2379
> 2020/11/10 08:41:42 control.go:350: [info] begin to set up database on  127.0.0.1:4000
> 2020/11/10 08:41:42 control.go:350: [info] begin to set up database on  127.0.0.1:20160
> 2020/11/10 08:41:42 control.go:350: [info] begin to set up database on  127.0.0.1:20161
> 2020/11/10 08:41:42 control.go:350: [info] begin to set up database on  127.0.0.1:20162
> 2020/11/10 08:41:42 control.go:371: [info] begin to set up client
> 2020/11/10 08:41:42 control.go:374: [info] begin to set up db client for node  127.0.0.1:4000
> 2020/11/10 08:41:42 panic_check.go:34: [warning] plugin panic check won't work
> 2020/11/10 08:41:42 control.go:298: [info] run client 0...
> 2020/11/10 08:41:45 leader_shuffle.go:269: [info] [leader shuffler] [leader=4] Merge leader region #126 and #2
> 2020/11/10 08:41:46 pipeline.go:264: [fatal] fail to lock: <nil>
> ```

It means a txn is not closed. I should have fixed it...

**Comment by youjiali1995** [comment]
> > I run this test case in local and it reports the following error
> > ```
> > ~/c/tipocket youjiali1995-pipelined-pessimistic-locking* > bin/pipelined-locking -tidb-server 127.0.0.1:4000 -pd-server 127.0.0.1:2379 -tikv-server 127.0.0.1:20160 -tikv-server 127.0.0.1:20161 -tikv-server 127.0.0.1:20162 -run-time 100s  -local-mode=true -strict=true
> > 2020/11/10 08:41:42 fixture.go:237: [info] Git Commit Hash: 6fd298ac05a546e6e325f167a342d28a8b320fb9
> > 2020/11/10 08:41:42 fixture.go:238: [info] UTC Build Time: 2020-11-10 08:29:02
> > W1110 08:41:42.012155   14314 client_config.go:541] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
> > W1110 08:41:42.012193   14314 client_config.go:546] error creating inClusterConfig, falling back to default config: unable to load in-cluster configuration, KUBERNETES_SERVICE_HOST and KUBERNETES_SERVICE_PORT must be defined
> > 2020/11/10 08:41:42 client.go:109: [warning] build config failed: invalid configuration: no configuration has been provided
> > 2020/11/10 08:41:42 client.go:47: [warning] error creating kube-client: must provide non-nil rest.Config to client.New
> > 2020/11/10 08:41:42 suit.go:87: [info] deploy cluster success, node:[ 127.0.0.1:4000  127.0.0.1:20160  127.0.0.1:20161  127.0.0.1:20162  127.0.0.1:2379], client node:[ 127.0.0.1:4000]
> > 2020/11/10 08:41:42 control.go:99: [info] start controller with &{Mode:Self scheduled mode DB: Nodes:[ 127.0.0.1:4000  127.0.0.1:20160  127.0.0.1:20161  127.0.0.1:20162  127.0.0.1:2379] ClientNodes:[ 127.0.0.1:4000] ClientCount:1 ChaosNamespace: RunRound:1 RunTime:1m40s RequestCount:0 History: ClientConfig:<nil>}
> > 2020/11/10 08:41:42 control.go:348: [info] begin to set up database
> > 2020/11/10 08:41:42 control.go:350: [info] begin to set up database on  127.0.0.1:2379
> > 2020/11/10 08:41:42 control.go:350: [info] begin to set up database on  127.0.0.1:4000
> > 2020/11/10 08:41:42 control.go:350: [info] begin to set up database on  127.0.0.1:20160
> > 2020/11/10 08:41:42 control.go:350: [info] begin to set up database on  127.0.0.1:20161
> > 2020/11/10 08:41:42 control.go:350: [info] begin to set up database on  127.0.0.1:20162
> > 2020/11/10 08:41:42 control.go:371: [info] begin to set up client
> > 2020/11/10 08:41:42 control.go:374: [info] begin to set up db client for node  127.0.0.1:4000
> > 2020/11/10 08:41:42 panic_check.go:34: [warning] plugin panic check won't work
> > 2020/11/10 08:41:42 control.go:298: [info] run client 0...
> > 2020/11/10 08:41:45 leader_shuffle.go:269: [info] [leader shuffler] [leader=4] Merge leader region #126 and #2
> > 2020/11/10 08:41:46 pipeline.go:264: [fatal] fail to lock: <nil>
> > ```
> 
> It means a txn is not closed. I should have fixed it...

I see. It means the key is locked. I think you ran this test in v4.x? @cosven

File: tests/pessimistic/pipeline/pipeline.go (hunk 3 above)
Comment (youjiali1995) [reply tests/pessimistic/pipeline/pipeline.go:177]: The table contains 3 regions when running 5 minutes.
```
+-----------+---------------------------+-----------+-----------+-----------------+---------------+------------+---------------+------------+----------------------+------------------+
| REGION_ID | START_KEY                 | END_KEY   | LEADER_ID | LEADER_STORE_ID | PEERS         | SCATTERING | WRITTEN_BYTES | READ_BYTES | APPROXIMATE_SIZE(MB) | APPROXIMATE_KEYS |
+-----------+---------------------------+-----------+-----------+-----------------+---------------+------------+---------------+------------+----------------------+------------------+
| 463       | t_56_5f728000000000000048 | t_59_r_0  | 737       | 5               | 622, 715, 737 | 0          | 269           | 0          | 2                    | 40960            |
| 440       | t_59_r_0                  | t_59_r_34 | 624       | 5               | 477, 624, 673 | 0          | 0             | 0          | 11                   | 164822           |
| 39        | t_59_r_34                 |           | 546       | 5               | 546, 703, 751 | 0          | 4395          | 0          | 35                   | 354064           |
+-----------+---------------------------+-----------+-----------+-----------------+---------------+------------+---------------+------------+----------------------+------------------+
```

File: tests/pessimistic/pipeline/pipeline.go (hunk 2 above)
Comment (youjiali1995) [reply tests/pessimistic/pipeline/pipeline.go:255]: It shouldn't panic if we don't kill instances. It's possible a region is unavailable due to the change in raftstore. If don't panic here, we can't notice it.

File: argo/workflow/pipelined-pessimistic-locking.yaml (hunk 4 above)
Comment (youjiali1995) [reply argo/workflow/pipelined-pessimistic-locking.yaml:17]: This test focuses on the success rate of pipelined locking in a normal environment. If enables `random kill`, some transaction can fail and we can't know whether it's caused by pipelined locking or node failure.

Comment (youjiali1995) [reply argo/workflow/pipelined-pessimistic-locking.yaml:17]: I think `shuffle region scheduler` is covered by `scatter table`?

**Review by cosven — COMMENTED** [review]
rest LGTM.

I'll try to run this case on K8s.

File: tests/pessimistic/pipeline/pipeline.go [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+		c.dbAddr = "127.0.0.1:4000"
+		c.dbStatusAddr = "127.0.0.1:10080"
+		c.pdAddr = "127.0.0.1:2379"
+		c.kvStatusAddrs = []string{"127.0.0.1:20180", "127.0.0.1:20181", "127.0.0.1:20182", "127.0.0.1:20183"}
+	} else {
+		c.dbAddr = dbNode.Address()
+		c.dbStatusAddr = fmt.Sprintf("%s-tidb.%s.svc:10080", dbNode.ClusterName, dbNode.Namespace)
+		c.pdAddr = fmt.Sprintf("%s-pd.%s.svc:2379", pdNode.ClusterName, pdNode.Namespace)
+		for _, kvNode := range kvNodes {
+			c.kvStatusAddrs = append(c.kvStatusAddrs, fmt.Sprintf("%s-tikv.%s.svc:20180", kvNode.ClusterName, kvNode.Namespace))
```
Comment (cosven) [inline tests/pessimistic/pipeline/pipeline.go:153]: The status addr is wrong
```
2020/11/17 02:24:11 control.go:298: [info] run client 0...
2020/11/17 02:24:11 pipeline.go:232: [fatal] failed to enable failpoints: Put "http://tipocket-pipelined-pessimistic-locking-tikv.tipocket-pipelined-pessimistic-locking.svc:20180/fail/apply_before_split": dial tcp: lookup tipocket-pipelined-pessimistic-locking-tikv.tipocket-pipelined-pessimistic-locking.svc on 10.96.0.10:53: no such host
```

one status addr example
```
curl -X POST http://tipocket-pipelined-pessimistic-locking-tikv-3.tipocket-pipelined-pessimistic-locking-tikv-pe
er.tipocket-pipelined-pessimistic-locking.svc:20180/fail/apply_before_split
```

**Comment by cosven** [comment]
After running for about 4 minutes, the following error occured
```
2020/11/19 10:20:52 pipeline.go:93: [warning] fail to schedule: POST request "http://tpctl-pipelined-locking-pd.tpctl-pipelined-locking.svc:2379/pd/api/v1/operators", got 500 "failed to add operator, maybe already have one"
2020/11/19 10:20:53 leader_shuffle.go:167: [info] [leader shuffler] [leader=6] Transfer leader region #198 from [6 7 5] to [7 1 4]
2020/11/19 10:20:53 pipeline.go:93: [warning] fail to schedule: POST request "http://tpctl-pipelined-locking-pd.tpctl-pipelined-locking.svc:2379/pd/api/v1/operators", got 500 "failed to add operator, maybe already have one"
2020/11/19 10:21:03 leader_shuffle.go:117: [info] [leader shuffler] [leader=6] Transfer leader from 6 to 5
2020/11/19 10:21:03 pipeline.go:370: [info] [1m10s] tps: 72.00 lock_err/s: 0.00 commit_err/s: 0.00 lock_err/commit_err/total: 0/0/22750(0.00%)
2020/11/19 10:21:04 leader_shuffle.go:286: [info] [leader shuffler] [leader=5] Split leader region #198
2020/11/19 10:21:54 pipeline.go:316: [fatal] fail to update: Error 1205: Lock wait timeout exceeded; try restarting transaction
2020/11/19 10:21:54 pipeline.go:316: [fatal] fail to update: Error 1205: Lock wait timeout exceeded; try restarting transaction
2020/11/19 10:21:54 pipeline.go:316: [fatal] fail to update: Error 1205: Lock wait timeout exceeded; try restarting transaction
```

**Comment by cosven** [comment]
The case will panic when the `-run-time` exceeded.
```
2020/11/19 10:43:54 leader_shuffle.go:286: [info] [leader shuffler] [leader=5] Split leader region #58
2020/11/19 10:43:57 pipeline.go:316: [fatal] fail to update: context deadline exceeded
```

**Comment by youjiali1995** [comment]
> UPDATED: Tipocket start 5 clients by default, so there maybe multiple contender for one row even if we set contender to 1. I will run this test case again.

How to control it?

**Comment by youjiali1995** [comment]
> The case will panic when the `-run-time` exceeded.
> 
> ```
> 2020/11/19 10:43:54 leader_shuffle.go:286: [info] [leader shuffler] [leader=5] Split leader region #58
> 2020/11/19 10:43:57 pipeline.go:316: [fatal] fail to update: context deadline exceeded
> ```

Fixed. Replace the root context by a background context in some functions. All SQLs can't run for a long time so I don't set a timeout for them.

**Review by cosven — APPROVED** [review]
LGTM

**Pull request merged**